File: TestParser/TestStr.fs
module TestParser.TestStr

open NUnit.Framework
open CbStyles.Cbon.Parser
open CbStyles.Cbon.Parser.Parser
open CbStyles.Parser

[<SetUp>]
let Setup () =
    ()

[<Test>]
let TestStr1 () =
    let code = Reader.reader "'asd'"
    let r = str code
    printf "%s" (r.ToString())
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), "asd"), r)

[<Test>]
let TestStr2 () =
    let code = Reader.reader "\"asd\""
    let r = str code
    printf "%s" (r.ToString())
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), "asd"), r)

[<Test>]
let TestStr3 () =
    let code = Reader.reader "'a\nd'"
    let r = str code
    printf "%s" (r.ToString())
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), "a\nd"), r)

[<Test>]
let TestStrEscape1 () =
    let code = Reader.reader "'a\\nd'"
    let r = str code
    printf "%s" (r.ToString())
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), "a\nd"), r)

[<Test>]
let TestStrEscape2 () =
    let code = Reader.reader "'a\\'d'"
    let r = str code
    printf "%s" (r.ToString())
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), "a'd"), r)

[<Test>]
let TestStrEscape3 () =
    let code = Reader.reader "'a\\u2A5F'"
    let r = str code
    printf "%s" (r.ToString())
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), "a\u2a5f"), r)

[<Test>]
let TestStrEscape4 () =
    let code = Reader.reader "'a\\u{2A5F}'"
    let r = str code
    printf "%s" (r.ToString())
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), "a\u2a5f"), r)

[<Test>]
let TestStrEscape5 () =
    let code = Reader.reader "'a\\u{u}'"
    let f () = str code |> ignore
    let e = Assert.Throws<ParserError> (new TestDelegate(f))
    printf "%s" (e.Message)
    ()

[<Test>]
let TestStrEscape6 () =
    let code = Reader.reader "'a\\u{}'"
    let f () = str code |> ignore
    let e = Assert.Throws<ParserError> (new TestDelegate(f))
    printf "%s" (e.Message)
    ()
    
[<Test>]
let TestStrEscape7 () =
    let code = Reader.reader "'a\\u{'"
    let f () = str code |> ignore
    let e = Assert.Throws<ParserError> (new TestDelegate(f))
    printf "%s" (e.Message)
    ()

File: TestParser/TestStr.cs
using NUnit.Framework;
using CbStyles.Cbon.Parser;

namespace TestParser
{
    public class TestStr
    {
        [SetUp]
        public void Setup() { }

        [Test]
        public void TestStr1()
        {
            var code = "'asd'";
            Parser.RunInReader(code, code => {
                var r = Parser.Str(code).Value.Item2;
                Assert.AreEqual("asd", r);
            });
        }

        [Test]
        public void TestStr2()
        {
            var code = "\"asd\"";
            Parser.RunInReader(code, code => {
                var r = Parser.Str(code).Value.Item2;
                Assert.AreEqual("asd", r);
            });
        }

        [Test]
        public void TestStr3()
        {
            var code = "'a\nd'";
            Parser.RunInReader(code, code => {
                var r = Parser.Str(code).Value.Item2;
                Assert.AreEqual("a\nd", r);
            });
        }

        [Test]
        public void TestStrEscape1()
        {
            var code = "'a\\nd'";
            Parser.RunInReader(code, code => {
                var r = Parser.Str(code).Value.Item2;
                Assert.AreEqual("a\nd", r);
            });
        }

        [Test]
        public void TestStrEscape2()
        {
            var code = "'a\\'d'";
            Parser.RunInReader(code, code => {
                var r = Parser.Str(code).Value.Item2;
                Assert.AreEqual("a'd", r);
            });
        }

        [Test]
        public void TestStrEscape3()
        {
            var code = "'\\u2A5F'";
            Parser.RunInReader(code, code => {
                var r = Parser.Str(code).Value.Item2;
                Assert.AreEqual("\u2a5f", r);
            });
        }

        [Test]
        public void TestStrEscape4()
        {
            var code = "'\\u{2A5F}'";
            Parser.RunInReader(code, code => {
                var r = Parser.Str(code).Value.Item2;
                Assert.AreEqual("\u2a5f", r);
            });
        }

        [Test]
        public void TestStrEscape5()
        {
            var code = "'\\u{u}'";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.Str(code);
            }));
            Assert.AreEqual("Illegal Unicode escape \t at 1:4", e.Message);
        }

        [Test]
        public void TestStrEscape6()
        {
            var code = "'\\u{}'";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.Str(code);
            }));
            Assert.AreEqual("Illegal Unicode escape \t at 1:4", e.Message);
        }

        [Test]
        public void TestStrEscape7()
        {
            var code = "'\\u{'";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.Str(code);
            }));
            Assert.AreEqual("Unicode escape not close or illegal characters \t at 1:4", e.Message);
        }
    }
}
